pkg/security/probe/process_killer.go

```diff
@@ -221,67 +283,84 @@ func (p *ProcessKiller) KillAndReport(kill *rules.KillDefinition, rule *rules.Ru
 			if !disarmer.allow(disarmer.executableCache, executable, func() {
 				disarmer.disarmedCount[executableDisarmerType]++
 				seclog.Warnf("disarmed kill action of rule `%s` because more than %d different executables triggered it in the last %s", rule.ID, disarmer.executable.capacity, disarmer.executable.period)
+				p.updateKillQueueAlarm(disarmer)
 			}) {
 				onActionBlockedByDisarmer(executableDisarmerType)
 				return false
 			}
 		}
 	}
 
-	pids, paths, err := p.getProcesses(scope, ev, entry)
+	pcs, err := p.getProcesses(scope, ev, entry)
 	if err != nil {
 		log.Errorf("unable to kill: %s", err)
 		return false
 	}
 
 	// if one pids is not allowed don't kill anything
-	if killAllowed, err := p.isKillAllowed(pids, paths); !killAllowed {
+	if killAllowed, err := p.isKillAllowed(pcs); !killAllowed {
 		log.Warnf("unable to kill: %v", err)
 		return false
 	}
 
 	sig := model.SignalConstants[kill.Signal]
 
-	var processesKilled int64
-	killedAt := time.Now()
-	for _, pid := range pids {
-		log.Debugf("requesting signal %s to be sent to %d", kill.Signal, pid)
-
-		if err := killFnc(uint32(pid), uint32(sig)); err != nil {
-			seclog.Debugf("failed to kill process %d: %s", pid, err)
-		} else {
-			processesKilled++
-		}
-	}
-
-	p.perRuleStatsLock.Lock()
-	var stats *processKillerStats
-	if stats = p.perRuleStats[rule.ID]; stats == nil {
-		stats = &processKillerStats{}
-		p.perRuleStats[rule.ID] = stats
-	}
-	stats.processesKilled += processesKilled
-	p.perRuleStatsLock.Unlock()
-
-	p.Lock()
-	defer p.Unlock()
-
 	report := &KillActionReport{
 		Scope:      scope,
 		Signal:     kill.Signal,
-		Status:     KillActionStatusPerformed,
 		CreatedAt:  ev.ProcessContext.ExecTime,
 		DetectedAt: ev.ResolveEventTime(),
-		KilledAt:   killedAt,
 		Pid:        ev.ProcessContext.Pid,
 		rule:       rule,
 	}
+
+	// TODO: merge periods to keep only one?
+	if disarmer != nil && p.FirstPeriodEnqueued(disarmer, sig, pcs, max(disarmer.container.period, disarmer.executable.period)) {
+		log.Warnf("rule %s triggered on first period, putting pids to kill on wait list", rule.ID)
+		report.Status = KillActionStatusQueued
+		report.resolved = true
```

Comment:
Not sure to understand why we mark it as resolved, is it not going to send directly the event without waiting for the action report ?

Reply:
It will yes. I've still to find a better way to handle reports for queued kills on rules with long periods. We can discuss offline